File: architecture-skills/design-patterns/examples/template-method/template-method.js
// Template Method pattern in JavaScript.
//
// The class form relies on subclasses overriding methods the base class
// calls. In practice JS often prefers a higher-order function that takes
// the varying steps as an object of functions — both forms are shown.

class SalesReport {
  // The template method: iteration and totalling never vary.
  generate(items) {
    const lines = [this.header()];
    let total = 0;
    for (const item of items) {
      const subtotal = item.units * item.price;
      lines.push(this.row(item, subtotal));
      total += subtotal;
    }
    lines.push(this.footer(total));
    return lines.join("\n");
  }

  header() {
    throw new Error("subclass must implement header()");
  }

  row(item, subtotal) {
    throw new Error("subclass must implement row()");
  }

  footer(total) {
    throw new Error("subclass must implement footer()");
  }
}

class CsvReport extends SalesReport {
  header() {
    return "product,units,price,subtotal";
  }

  row(item, subtotal) {
    return `${item.product},${item.units},${item.price.toFixed(2)},${subtotal.toFixed(2)}`;
  }

  footer(total) {
    return `total,,,${total.toFixed(2)}`;
  }
}

// --- Alternative: composition with a steps object (often preferred) ---

function generateReport(steps, items) {
  const lines = [steps.header()];
  let total = 0;
  for (const item of items) {
    const subtotal = item.units * item.price;
    lines.push(steps.row(item, subtotal));
    total += subtotal;
  }
  lines.push(steps.footer(total));
  return lines.join("\n");
}

const markdownSteps = {
  header: () => "| Product | Units | Subtotal |\n| --- | --- | --- |",
  row: (item, subtotal) => `| ${item.product} | ${item.units} | $${subtotal.toFixed(2)} |`,
  footer: (total) => `\n**Total: $${total.toFixed(2)}**`,
};

const items = [
  { product: "Keyboard", units: 2, price: 80.0 },
  { product: "Mouse", units: 1, price: 25.5 },
  { product: "Monitor", units: 3, price: 210.0 },
];

console.log("--- CsvReport (inheritance) ---");
console.log(new CsvReport().generate(items));
console.log();
console.log("--- MarkdownReport (composition) ---");
console.log(generateReport(markdownSteps, items));

File: architecture-skills/design-patterns/examples/template-method/template-method.go
// Template Method pattern in Go.
//
// Go has no inheritance, so the idiomatic form is composition: the
// template is a plain function that fixes the skeleton and takes an
// interface supplying the varying steps. This avoids the classic
// pattern's tight base-class coupling entirely.

package main

import "fmt"

type LineItem struct {
	Product string
	Units   int
	Price   float64
}

func (l LineItem) Subtotal() float64 {
	return float64(l.Units) * l.Price
}

// ReportSteps supplies only the varying steps.
type ReportSteps interface {
	Header() string
	Row(item LineItem) string
	Footer(total float64) string
}

// GenerateReport is the template method: iteration and totalling never vary.
func GenerateReport(steps ReportSteps, items []LineItem) string {
	out := steps.Header() + "\n"
	total := 0.0
	for _, item := range items {
		out += steps.Row(item) + "\n"
		total += item.Subtotal()
	}
	return out + steps.Footer(total)
}

type CsvReport struct{}

func (CsvReport) Header() string {
	return "product,units,price,subtotal"
}

func (CsvReport) Row(item LineItem) string {
	return fmt.Sprintf("%s,%d,%.2f,%.2f", item.Product, item.Units, item.Price, item.Subtotal())
}

func (CsvReport) Footer(total float64) string {
	return fmt.Sprintf("total,,,%.2f", total)
}

type MarkdownReport struct{}

func (MarkdownReport) Header() string {
	return "| Product | Units | Subtotal |\n| --- | --- | --- |"
}

func (MarkdownReport) Row(item LineItem) string {
	return fmt.Sprintf("| %s | %d | $%.2f |", item.Product, item.Units, item.Subtotal())
}

func (MarkdownReport) Footer(total float64) string {
	return fmt.Sprintf("\n**Total: $%.2f**", total)
}

func main() {
	items := []LineItem{
		{"Keyboard", 2, 80.00},
		{"Mouse", 1, 25.50},
		{"Monitor", 3, 210.00},
	}

	reports := []struct {
		name  string
		steps ReportSteps
	}{
		{"CsvReport", CsvReport{}},
		{"MarkdownReport", MarkdownReport{}},
	}

	for _, r := range reports {
		fmt.Printf("--- %s ---\n%s\n\n", r.name, GenerateReport(r.steps, items))
	}
}
